### DISCUSSION TITLE
Hierarchical Layout Algorithm

### DISCUSSION THREAD

**Discussion by atillaaliyev** [clientIO/joint#2753] [Q&A]
### Introduction

Hi, 
I will make project about EOD tree (End of days Job) . This EOD tree is watched by a team everyday. There are a lot of jobs that are related each other. I want to make this;
[EOD09062021.pdf](https://github.com/user-attachments/files/16941741/EOD09062021.pdf)

I will give nodes and edges and the library will export meaningful output like above. I have found some library like this;  https://www.yworks.com/demos/showcase/layoutstyles/?layout=hierarchic&sample=hierarchic

However, I want to learn, can I make this with JOINT. I have reviewed all demos but i did not find.
Thanks,

### Does your question relate to JointJS or JointJS+. Select both if applicable.

JointJS, JointJS+

**Comment by kumilingus** [comment]
If you need parallel orthogonal links, you will need to use a 3rd party library to position the nodes and route the links:
- **MSAGLJS**: https://microsoft.github.io/msagljs/
- **ELKJS**:  https://github.com/kieler/elkjs

*MSAGL example:*
https://codepen.io/jointjs/pen/zYQQZpg/c2f18bb13f4bbd41c4ade3a8b50540ef
<img width="515" alt="image" src="https://github.com/user-attachments/assets/311a984f-21fe-4a7b-8c27-837e6a354a7b">

*ELK examples:*
https://www.jointjs.com/demos/eclipse-kernel-layout
<img width="281" alt="elkjs2" src="https://github.com/user-attachments/assets/b4aab02e-9e2d-4955-ba19-739baed2334d">
<img width="1068" alt="elkjs1" src="https://github.com/user-attachments/assets/1311d048-f188-4d8e-8314-c25b218a24bd">

**Reply by atillaaliyev** [reply]
Thank you @kumilingus,

Neither msagl nor elk don't give me that I want output. The nodes are calculated far to each other by these. However they worked good. Could you offer me library that calculates the best coordinates? My company want to work with JointJs+. However, they expect good layout like yfiles. Obviously, own algorithm of yfiles calculates coordinates better than elk and msagl.

**Reply by kumilingus** [reply]
Could you please explain why you think these two layouts do not meet your expectations? Perhaps there is a configuration that can be used to modify the output (some screenshots would be awesome).

Note that there is no such thing as the best coordinates. Each use case / application may require a different layout.

I don't know of any other library that I can recommend that can be compared to the ones already mentioned.

**Reply by atillaaliyev** [reply]
Hi @kumilingus ,
Sorry my late answer.
I share elk ve msagl example for my case;

https://codesandbox.io/p/sandbox/elk-msagl-example-v74pg6

And this is yfiles screenshot. It is more clear and readable. I talked about this algorithm.
![image](https://github.com/user-attachments/assets/b528fcce-6a88-4b7a-a57c-5dd1e99fccbf)

**Reply by kumilingus** [reply]
Thank you for the live examples.

The links from the _yFiles_ examples are not orthogonal. If this is not a requirement, you can also look at the [DirectedGraph](https://docs.jointjs.com/learn/features/automatic-layouts/directed-graph) layout, have you looked?

**Reply by atillaaliyev** [reply]
Yes, I had tried before. However, the result was unsuccessful.

https://codesandbox.io/p/sandbox/winter-browser-ft9z8c

**Reply by kumilingus** [reply]
This is the result of DirectedGraph for reference (after I moved the link anchors to the bottom/top):

https://codesandbox.io/p/sandbox/great-mirzakhani-ddd4wv

<img width="2483" alt="image" src="https://github.com/user-attachments/assets/6d4feecb-7705-454b-9fc0-5862fb9a2df4">

If this doesn't work for you either (and both examples use the same dataset), I can't recommend any other open-source library at this time.
